nexus/db-queries/src/db/datastore/snapshot.rs

```diff
@@ -48,127 +44,49 @@ impl DataStore {
         let gen = snapshot.gen;
         opctx.authorize(authz::Action::CreateChild, authz_project).await?;
 
-        #[derive(Debug, thiserror::Error)]
-        pub enum CustomError {
-            #[error("Resource already exists")]
-            ResourceAlreadyExists,
-
-            #[error("saw AsyncInsertError")]
-            InsertError(AsyncInsertError),
-        }
-
-        type TxnError = TransactionError<CustomError>;
+        use db::schema::snapshot::dsl;
+        use diesel::query_dsl::methods::FilterDsl;
 
-        let snapshot_name = snapshot.name().to_string();
         let project_id = snapshot.project_id;
+        let snapshot_name = snapshot.name().to_string();
+        let snapshot_id = snapshot.id();
+
+        let mut snapshots: Vec<Snapshot> = Project::insert_resource(
+            project_id,
+            diesel::insert_into(dsl::snapshot)
+                .values(snapshot)
+                .on_conflict((dsl::project_id, dsl::name))
+                .filter_target(dsl::time_deleted.is_null())
+                .do_update()
+                .set(dsl::time_modified.eq(dsl::time_modified))
```

Comment:
Do we still want to update the old record in this case?

Reply:
Definitely not, but I'm at a loss as how to do that.

Reply:
Can we use `on_conflict().do_nothing()`?

Reply:
If I do

```rust
        let mut snapshots: Vec<Snapshot> = Project::insert_resource(
            project_id,
            diesel::insert_into(dsl::snapshot)
                .values(snapshot)
                .on_conflict((dsl::project_id, dsl::name))
                .filter_target(dsl::time_deleted.is_null())
                .do_nothing()
                .filter(dsl::id.eq(snapshot_id)),
        )
        .insert_and_get_results_async(self.pool_authorized(opctx).await?)
```

then I see a syntax error:

```
thread 'integration_tests::snapshots::test_create_snapshot_record_idempotent' panicked at 'called `Result::unwrap()` on an `Err` value: InternalError { internal_message: "unexpected database error: at or near \"where\": syntax error" }', nexus/tests/integration_tests/snapshots.rs:996:10
```

If I remove the `.filter(dsl::id.eq(snapshot_id)),` on the end of that (so it's just `on_conflict + do_nothing`, I see `test_create_snapshot_record_idempotent` failing on the second call here:

```rust
    // Test project_ensure_snapshot is idempotent

    let snapshot_created_1 = datastore
        .project_ensure_snapshot(&opctx, &authz_project, snapshot.clone())
        .await
        .unwrap();
                
    let snapshot_created_2 = datastore
        .project_ensure_snapshot(&opctx, &authz_project, snapshot.clone())
        .await
        .unwrap();
```

saying the snapshot already exists:

```
thread 'integration_tests::snapshots::test_create_snapshot_record_idempotent' panicked at 'called `Result::unwrap()` on an `Err` value: ObjectAlreadyExists { type_name: Snapshot, object_name: "snapshot" }', nexus/tests/integration_tests/snapshots.rs:1001:10
```